tests/unit/sproxyd.js

```diff
@@ -297,14 +307,70 @@ const clientImmutableWithFailover = new Sproxy({
             const upStream = new stream.PassThrough();
             upStream.write(upload.slice(0, upload.length - 10));
             setTimeout(() => upStream.destroy(), 500);
-            client.put(upStream, upload.length, parameters, reqUid,
-                err => {
+            async.series([
+                next => {
+                    client.put(upStream, upload.length, parameters, reqUid,
+                        err => {
+                            if (err) {
+                                next();
+                            } else {
+                                assert.fail('expected an immediate error from sproxyd');
+                            }
+                        });
+                },
+                // Next request should go wel
+                next => {
+                    const upStream = new stream.PassThrough();
+                    upStream.write(upload);
+                    client.put(upStream, upload.length, parameters, reqUid,
+                        (err, key) => {
+                            if (err) {
+                                done(err);
```

Comment:
```suggestion
                                next(err);
```

Reply:
I think the point is that `async.series` is (and should be) a blackbox : whose semantics is that each step MUST call `next()`, and `done()` would thus only be called by the "end" callback:

```
async.series([
   next => { .... next() },
   next => { .... next() },
   next => { .... next() },
], done)
```

so to follow the pattern properly, indeed you should simply call `next(err)` but also modify the next (and last) step to pass `next` instead `done` from `checkKeyContent`...

This hides the "location" of the call to done indeed, but anyway backtraces are not really used to report the error : the relevant information should be passed through the error itself.

---

That said, not sure there is really a _practical_ problem with bypassing the call to next in `async.series()`, maybe @jonathan-gramain has more insight ?

Reply:
Agree with @francoisferrand , it's more to follow the general pattern. For failing a test it should not matter, but code in general should make sure to always call the individual task callbacks (`next`) until the final control loop callback (`done`) is called. One reason could be to give an opportunity for the control flow logic to free up resources when calling `next(err)` or at the last task's `next()` call. (I don't think `async.series` has anything to free but it could be the case for others, and it's up to the implementation to decide what to do when it's done with the execution).

By the way, I think the error backtrace would still contain the exact location of the error, because `async.series` does't call `process.nextTick` before it calls the final callback, it just calls it synchronously.

And, indeed you'd have to catch the error at https://github.com/scality/sproxydclient/pull/305/files#diff-b186714b12094c1a4c22bde8a30b6872b181c2a4023faf4fad76d901f73c2012R336, it could be done with a one-liner `assert.ifError(err)`.

Reply:
Thanks for the insights! I fixed it.